src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -307,44 +336,118 @@ private void EventLoop()
+                    }
+                    else if (_handleToContextMap.TryGetValue(socketEvent.Data, out SocketAsyncContext context))
                     {
-                        IntPtr handle = _buffer[i].Data;
-                        if (handle == ShutdownHandle)
+                        if ((socketEvent.Events & Interop.Sys.SocketEvents.Error) != 0)
                         {
-                            shutdown = true;
+                            // there was an error, we use the non-buffered execution path (this should be very rare)
+                            context.HandleEvents(socketEvent.Events);
                         }
                         else
                         {
-                            Debug.Assert(handle.ToInt64() < MaxHandles.ToInt64(), $"Unexpected values: handle={handle}, MaxHandles={MaxHandles}");
-                            _handleToContextMap.TryGetValue(handle, out SocketAsyncContext context);
-                            if (context != null)
-                            {
-                                context.HandleEvents(_buffer[i].Events);
-                                context = null;
-                            }
+                            context.AddWaitingOperationsToBatch(socketEvent.Events, ioControlBlocks, batchedOperations, ref batchedCount);
                         }
                     }
                 }
 
-                FreeNativeResources();
-            }
-            catch (Exception e)
-            {
-                Environment.FailFast("Exception thrown from SocketAsyncEngine event loop: " + e.ToString(), e);
+                if (batchedCount > 0)
+                {
+                    int result = Interop.Sys.IoSubmit(_aioContext.Ring, batchedCount, _aioBlocksPointers);
+                    if (result != batchedCount)
+                    {
+                        Interop.Error lastError = Interop.Sys.GetLastError();
+                        throw new InternalException($"IoSubmit has failed with {lastError}, returned {result} instead {batchedCount}"); // todo: implement a while loop
+                    }
+
+                    // todo: perf: avoid the syscall by using a well known pattern that reads from the ring
+                    result = Interop.Sys.IoGetEvents(_aioContext.Ring, batchedCount, batchedCount, _aioEvents);
+                    if (result != batchedCount)
+                    {
+                        Interop.Error lastError = Interop.Sys.GetLastError();
+                        throw new InternalException($"IoGetEvents has failed with {lastError}, returned {result} instead {batchedCount}");
+                    }
+
+                    ReadOnlySpan<Interop.Sys.IoEvent> events = new ReadOnlySpan<Interop.Sys.IoEvent>(_aioEvents, batchedCount);
+                    for (int i = 0; i < events.Length; i++)
+                    {
+                        batchedOperations[events[i].Data].HandleBatchEvent(in events[i]);
```

Comment:
Is there some code path that ensures other operations that are pending in the queue get executed?

Reply:
Tom, I am not sure if I understand. We batch what we can and dispatch the rest, then here we handle only what was batched.

Reply:
> dispatch the rest

We can't dispatch the rest. The operations need to be handled sequentially.

We also need to take care of new operations being added to the queue while the batched operations are executing. So at the end of `HandleBatchEvent` we must check the queue.